### PR TITLE
New FenderStrings

### PR DESCRIPTION
**Pull request by FuzzyNovaGoblin**
this change is not done yet it still needs to have some methods implemented for strings that were previously only implemented for lists, and there are a number of times where `toString` is called to use the rust std string functions but these methods can be implemented for the `FenderString` to eliminate those `toString` calls

### CODE REVIEW

File: src/type_sys/fender_value/fender_strings.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
+use std::{cell::RefCell, fmt::Display, hash::Hash, ops::Index};
+
+use crate::index_oob;
+
+#[derive(Debug, Clone)]
+pub struct FenderString {
+    chars: Vec<char>,
+    cacheable: bool,
```
Comment (boxbeam) [inline src/type_sys/fender_value/fender_strings.rs:8]: So is the point that you can create some strings that will never be cached? I'm not really sure why we need this field.

File: src/type_sys/fender_value/fender_strings.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            return false;
+        }
+        if self.cacheable {
+            *self.cached_string.get_mut() = None;
+        }
+        self.chars.swap(pos_a, pos_b);
+        true
+    }
+
+    pub fn insert_char(&mut self, index: usize, value: char) -> bool {
```
Comment (boxbeam) [inline src/type_sys/fender_value/fender_strings.rs:164]: With us introducing all of these functions, I would like to have at least a few unit tests for them. Could we create something like `ensure_valid` that makes sure the char vec and the cached string are identical, and then use that for assertions in tests?

**Review by boxbeam — CHANGES_REQUESTED** [review]
Since this is such a large changeset, could you try to summarize some of what you did? Some things in here seem unrelated to fender strings.

File: src/type_sys/fender_value/mod.rs [hunk 3]
```diff
@@ -421,7 +451,7 @@ impl ToString for FenderValue {
             FenderValue::Int(i) => i.to_string(),
             FenderValue::Float(f) => f.to_string(),
             FenderValue::Char(c) => c.to_string(),
-            FenderValue::String(s) => s.deref().clone(),
+            FenderValue::String(s) => s.iter().collect(),
```
Comment (boxbeam) [inline src/type_sys/fender_value/mod.rs:454]: Is there a way to get the cached string? For example, I think `print_func` still uses `to_string` and so would copy the whole string whenever it's printed. Maybe we need to have a way to convert a type into an `impl IntoIterator<Item = char>`?

**Comment by FuzzyNovaGoblin** [comment]
its not letting me comment on it but @Redempt said
> Since this is such a large changeset, could you try to summarize some of what you did? Some things in here seem unrelated to fender strings.

Can you give an example? I went through quickly and didn't see anything that wasn't just updating stuff to work with the new strings or the creation of the new strings themselves. I did see a macro for writing the out of bounds errors because there was just a lot of duplication and wanted to make sure they were all the same, but even that was because I was adjusting methods for the new strings.

**Comment by boxbeam** [comment]
Okay, I left that comment a while ago so I don't know what I was talking about specifically anymore.

**Pull request closed without merging**